### PR TITLE
Initial partest support

### PR DESCRIPTION
**Pull request by michaelpigg**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -202,6 +202,34 @@ lazy val forkjoin = configureAsForkOfJavaProject(project)
 
 lazy val asm = configureAsForkOfJavaProject(project)
 
+lazy val test = project.
+  settings(disableDocsAndPublishingTasks: _*).
+  settings(commonSettings: _*).
+  settings(
+    libraryDependencies += "org.scala-lang.modules" %% "scala-partest-interface" % "0.5.0-SNAPSHOT" % "test",
```
Comment (gkossakowski) [inline build.sbt]: Do we need to depend on SNAPSHOT version here?

File: build.sbt [hunk 2]
```diff
@@ -202,6 +202,34 @@ lazy val forkjoin = configureAsForkOfJavaProject(project)
...
+    libraryDependencies += "org.scala-lang.modules" %% "scala-partest" % "1.0.5" % "test",
+    libraryDependencies += "org.scalacheck" %% "scalacheck" % "1.11.4" % "test",
+    libraryDependencies += "org.scala-lang.modules" %% "scala-xml" % "1.0.3" % "test",
```
Comment (gkossakowski) [inline build.sbt]: We could collapse this to:

``` scala
libraryDependencies ++= Seq(
 "org.scala-lang.modules" %% "scala-partest" % "1.0.5" % "test",
 "org.scalacheck" %% "scalacheck" % "1.11.4" % "test",
 "org.scala-lang.modules" %% "scala-xml" % "1.0.3" % "test")
```

File: build.sbt (hunk 1 above)
Comment (michaelpigg) [inline build.sbt]: I don't think so, changing to 0.4.0.

File: build.sbt (hunk 2 above)
Comment (michaelpigg) [inline build.sbt]: Indeed. Fixed.

**Comment by gkossakowski** [comment]
Hey!

I tried running partest tests and I got:

```
[info] Partest version:     
[info] Compiler under test: $baseDir/compiler
[info] Scala version is:    Scala compiler version 2.11.6-SNAPSHOT-20150310-172550-76b18f6cf2 -- Copyright 2002-2013, LAMP/EPFL
[info] Scalac options are:  
[info] Compilation Path:    /Users/grek/scala/scala-master/target/test/test-classes:$baseDir/test:$baseDir/compiler:$baseDir/library:/Users/grek/scala/scala-master/build-sbt/libs/classes/forkjoin:$baseDir/reflect:/Users/grek/scala/scala-master/build-sbt/libs/classes/asm:$baseDir/actors:$baseDir/repl:$sourceDir/lib/annotations.jar:$sourceDir/lib/enums.jar:$sourceDir/lib/genericNest.jar:$sourceDir/lib/javac-artifacts.jar:$sourceDir/lib/jsoup-1.3.1.jar:$sourceDir/lib/macro210.jar:$sourceDir/lib/methvsfield.jar:$sourceDir/lib/nest.jar:$sourceDir/lib/scalacheck.jar:/Users/grek/scala/scala-master/test/lib:/Users/grek/.ivy2/cache/org.apache.ant/ant/jars/ant-1.9.4.jar:/Users/grek/.ivy2/cache/org.apache.ant/ant-launcher/jars/ant-launcher-1.9.4.jar:/Users/grek/.ivy2/cache/jline/jline/jars/jline-2.12.jar:/Users/grek/.ivy2/cache/org.scala-lang.modules/scala-partest-interface_2.11/jars/scala-partest-interface_2.11-0.4.0.jar:/Users/grek/.ivy2/cache/org.scala-sbt/test-interface/jars/test-interface-1.0.jar:/Users/grek/.ivy2/cache/org.scala-lang.modules/scala-partest_2.11/jars/scala-partest_2.11-1.0.5.jar:/Users/grek/.ivy2/cache/com.googlecode.java-diff-utils/diffutils/jars/diffutils-1.3.0.jar:/Users/grek/.ivy2/cache/org.scalacheck/scalacheck_2.11/jars/scalacheck_2.11-1.11.4.jar:/Users/grek/.ivy2/cache/org.scala-lang.modules/scala-parser-combinators_2.11/bundles/scala-parser-combinators_2.11-1.0.1.jar:/Users/grek/.ivy2/cache/org.scala-lang.modules/scala-xml_2.11/bundles/scala-xml_2.11-1.0.3.jar:/Users/grek/.sbt/boot/scala-2.10.4/org.scala-sbt/sbt/0.13.7/test-agent-0.13.7.jar:/Users/grek/.sbt/boot/scala-2.10.4/org.scala-sbt/sbt/0.13.7/test-interface-1.0.jar
[info] Java binaries in:    /System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home/bin
[info] Java runtime is:     Java HotSpot(TM) 64-Bit Server VM (build 20.65-b04-466.1, mixed mode)
[info] Java options are:    
[info] baseDir:             /Users/grek/scala/scala-master/build-sbt/quick/classes
[info] sourceDir:           /Users/grek/scala/scala-master/test/files
[info]     
!!  1 - scalacheck/t4147.scala                    [compilation failed]
!!  2 - scalacheck/treeset.scala                  [compilation failed]
[error] Uncaught exception when running partest: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Error running /Users/grek/scala/scala-master/test/files/scalacheck/array-new.scala
[trace] Stack trace suppressed: run last test/test:test for the full output.
[error] Error: Total 1, Failed 0, Errors 1, Passed 0
[error] Error during tests:
[error]     partest
[error] (test/test:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 83 s, completed Mar 10, 2015 5:36:17 PM
```

Is this what you see?

**Comment by michaelpigg** [comment]
No, for me it runs scalacheck, pos, neg pretty cleanly. run has a number of failures. all of instrumented fails. I am running Java 1.8.

**Comment by michaelpigg** [comment]
jvm was maybe 50% successful, res was clean. Then it blew up in scalap:

```
[error] Uncaught exception when running partest: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Error running /Users/michael/projects/mpigg-scala/test/files/scalap/abstractClass.scala
sbt.ForkMain$ForkError: java.lang.RuntimeException: Error running /Users/michael/projects/mpigg-scala/test/files/scalap/abstractClass.scala
    at java.util.concurrent.FutureTask.report(FutureTask.java:122)
    at java.util.concurrent.FutureTask.get(FutureTask.java:192)
    at scala.tools.partest.nest.SuiteRunner$$anonfun$runTestsForFiles$1.apply(Runner.scala:820)
    at scala.tools.partest.nest.SuiteRunner$$anonfun$runTestsForFiles$1.apply(Runner.scala:820)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
    at scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
    at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
    at scala.tools.partest.nest.SuiteRunner.runTestsForFiles(Runner.scala:820)
    at scala.tools.partest.nest.AntRunner.runSet(AntRunner.scala:36)
    at scala.tools.partest.nest.AntRunner$$anonfun$3.apply(AntRunner.scala:52)
    at scala.tools.partest.nest.AntRunner$$anonfun$3.apply(AntRunner.scala:52)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
    at scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
    at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
    at scala.tools.partest.nest.AntRunner.execute(AntRunner.scala:52)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:483)
    at scala.tools.partest.PartestTask.execute(TestingInterface.scala:75)
    at sbt.ForkMain$Run$2.call(ForkMain.java:294)
    at sbt.ForkMain$Run$2.call(ForkMain.java:284)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: sbt.ForkMain$ForkError: Error running /Users/michael/projects/mpigg-scala/test/files/scalap/abstractClass.scala
    at scala.tools.partest.nest.SuiteRunner.liftedTree1$1(Runner.scala:801)
    at scala.tools.partest.nest.SuiteRunner.runTest(Runner.scala:799)
    at scala.tools.partest.nest.SuiteRunner$$anonfun$21$$anonfun$apply$6.apply(Runner.scala:814)
    at scala.tools.partest.nest.SuiteRunner$$anonfun$21$$anonfun$apply$6.apply(Runner.scala:814)
    at scala.tools.partest.package$$anon$2.call(package.scala:133)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: sbt.ForkMain$ForkError: scala.tools.scalap.scalasig.ByteCode$
    at java.net.URLClassLoader$1.run(URLClassLoader.java:372)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.net.URLClassLoader.findClass(URLClassLoader.java:360)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
    at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
    at java.lang.Class.forName0(Native Method)
    at java.lang.Class.forName(Class.java:260)
    at scala.tools.partest.nest.Runner$$anonfun$20.apply(Runner.scala:706)
    at scala.tools.partest.nest.Runner$$anonfun$20.apply(Runner.scala:706)
    at scala.util.Try.getOrElse(Try.scala:78)
    at scala.tools.partest.nest.Runner.scala$tools$partest$nest$Runner$$decompileClass(Runner.scala:706)
    at scala.tools.partest.nest.Runner$$anonfun$runScalapTest$1.apply$mcZ$sp(Runner.scala:719)
    at scala.tools.partest.nest.Runner$$anonfun$runTestCommon$1.apply$mcZ$sp(Runner.scala:508)
    at scala.tools.partest.nest.Runner.runInContext(Runner.scala:397)
    at scala.tools.partest.nest.Runner.runTestCommon(Runner.scala:507)
    at scala.tools.partest.nest.Runner.runScalapTest(Runner.scala:715)
    at scala.tools.partest.nest.Runner.run(Runner.scala:681)
    at scala.tools.partest.nest.SuiteRunner$$anonfun$liftedTree1$1$1.apply(Runner.scala:799)
    at scala.tools.partest.nest.SuiteRunner$$anonfun$liftedTree1$1$1.apply(Runner.scala:799)
    at scala.tools.partest.package$.timed(package.scala:127)
    at scala.tools.partest.nest.SuiteRunner.liftedTree1$1(Runner.scala:799)
    at scala.tools.partest.nest.SuiteRunner.runTest(Runner.scala:799)
    at scala.tools.partest.nest.SuiteRunner$$anonfun$21$$anonfun$apply$6.apply(Runner.scala:814)
    at scala.tools.partest.nest.SuiteRunner$$anonfun$21$$anonfun$apply$6.apply(Runner.scala:814)
    at scala.tools.partest.package$$anon$2.call(package.scala:133)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
[error] Error: Total 1, Failed 0, Errors 1, Passed 0
[error] Error during tests:
[error]     partest
[error] (test/test:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 4390 s, completed Mar 10, 2015 9:13:44 PM
```

**Comment by michaelpigg** [comment]
Looks like pull-binary-libs.sh needs to be run in order to get the jars in test/files/lib.

**Comment by michaelpigg** [comment]
Currently have test failures in:
- instrumented - due to missing jar
- presentation - missing `scala.tools.nsc.interactive`
- run/repl-javap-app.scala - maybe Java 8?

All other tests are successful.

**Comment by michaelpigg** [comment]
Issue with run/repl-javap-app.scala is Java 8. Works okay compiling with Java 7. All other tests are now successful.

**Comment by adriaanm** [comment]
:fireworks:

**Comment by gkossakowski** [comment]
Hi @michaelpigg! Excellent news!

I got knocked down by a flu. I'll have a look into detail at your PR once I feel a bit better.

File: build.sbt [hunk 3]
```diff
@@ -202,10 +204,76 @@ lazy val forkjoin = configureAsForkOfJavaProject(project)
...
+      partestBuildDirectory.value / "pack/lib" / s"${resolvedArtifact.name}.${resolvedArtifact.extension}"
+    },
+    // add required manifest entry - previously included from file
+    packageOptions in (Compile, packageBin) += 
+      Package.ManifestAttributes( "Premain-Class" -> "scala.tools.partest.javaagent.ProfilingAgent" ),
+    // we need to build this to a JAR
+    exportJars := true
+  ).dependsOn(scalaAsm)
+
+lazy val scalaAsm = (project in file(".") / "src" / "asm").
```
Comment (gkossakowski) [inline build.sbt]: Why is this needed? The `asm` project defines that already.

**Comment by gkossakowski** [comment]
Hi!

I just tried running partest and I got:

```
> test
[warn] Scala version was updated by one of library dependencies:
[warn]  * org.scala-lang:scala-library:2.11.5 -> 2.11.6-SNAPSHOT
[warn] To force scalaVersion, add the following:
[warn]  ivyScala := ivyScala.value map { _.copy(overrideScalaVersion = true) }
[warn] Run 'evicted' to see detailed eviction warnings
[warn] No main class detected
[info] Partest version:     
[info] Compiler under test: $baseDir/compiler
[info] Scala version is:    Scala compiler version 2.11.6-SNAPSHOT-20150313-180153-1e20c21b07 -- Copyright 2002-2013, LAMP/EPFL
[info] Scalac options are:  
[info] Compilation Path:    /Users/grek/scala/scala-master/target/test/test-classes:$baseDir/test:$baseDir/compiler:$baseDir/library:/Users/grek/scala/scala-master/build-sbt/libs/classes/forkjoin:$baseDir/reflect:/Users/grek/scala/scala-master/build-sbt/libs/classes/asm:$baseDir/interactive:$baseDir/actors:$baseDir/repl:$baseDir/scalap:$baseDir/partest-extras:/Users/grek/scala/scala-master/build/pack/lib/scala-partest-javaagent.jar:$baseDir/scalaAsm:$baseDir/scaladoc:$sourceDir/lib/annotations.jar:$sourceDir/lib/enums.jar:$sourceDir/lib/genericNest.jar:$sourceDir/lib/javac-artifacts.jar:$sourceDir/lib/jsoup-1.3.1.jar:$sourceDir/lib/macro210.jar:$sourceDir/lib/methvsfield.jar:$sourceDir/lib/nest.jar:$sourceDir/lib/scalacheck.jar:/Users/grek/scala/scala-master/test/lib:/Users/grek/.ivy2/cache/org.apache.ant/ant/jars/ant-1.9.4.jar:/Users/grek/.ivy2/cache/org.apache.ant/ant-launcher/jars/ant-launcher-1.9.4.jar:/Users/grek/.ivy2/cache/jline/jline/jars/jline-2.12.1.jar:/Users/grek/.ivy2/cache/org.scala-lang.modules/scala-partest_2.11/jars/scala-partest_2.11-1.0.5.jar:/Users/grek/.ivy2/cache/com.googlecode.java-diff-utils/diffutils/jars/diffutils-1.3.0.jar:/Users/grek/.ivy2/cache/org.scala-sbt/test-interface/jars/test-interface-1.0.jar:/Users/grek/.ivy2/cache/org.scala-lang.modules/scala-xml_2.11/bundles/scala-xml_2.11-1.0.3.jar:/Users/grek/.ivy2/cache/org.scala-lang.modules/scala-parser-combinators_2.11/bundles/scala-parser-combinators_2.11-1.0.3.jar:/Users/grek/.ivy2/cache/org.scala-lang.modules/scala-partest-interface_2.11/jars/scala-partest-interface_2.11-0.5.0.jar:/Users/grek/.ivy2/cache/org.scalacheck/scalacheck_2.11/jars/scalacheck_2.11-1.11.4.jar:/Users/grek/.sbt/boot/scala-2.10.4/org.scala-sbt/sbt/0.13.7/test-agent-0.13.7.jar:/Users/grek/.sbt/boot/scala-2.10.4/org.scala-sbt/sbt/0.13.7/test-interface-1.0.jar
[info] Java binaries in:    /System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home/bin
[info] Java runtime is:     Java HotSpot(TM) 64-Bit Server VM (build 20.65-b04-466.1, mixed mode)
[info] Java options are:    
[info] baseDir:             /Users/grek/scala/scala-master/build-sbt/quick/classes
[info] sourceDir:           /Users/grek/scala/scala-master/test/files
[info]     
!!  1 - scalacheck/t4147.scala                    [compilation failed]
!!  2 - scalacheck/treeset.scala                  [compilation failed]
[error] Uncaught exception when running partest: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Error running /Users/grek/scala/scala-master/test/files/scalacheck/array-new.scala
[trace] Stack trace suppressed: run last test/test:test for the full output.
[error] Error: Total 1, Failed 0, Errors 1, Passed 0
[error] Error during tests:
[error]     partest
[error] (test/test:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 58 s, completed Mar 13, 2015 6:28:38 PM
```

@michaelpigg: are you sure you pushed all of your changes?
I pushed the branch I was using to test partest integration here: https://github.com/gkossakowski/scala/tree/partest-sbt-build
Let me know if I didn't mess up anything while merging.

**Comment by michaelpigg** [comment]
I'm not sure why you get that failure. I'll try your branch.

File: build.sbt (hunk 3 above)
Comment (michaelpigg) [inline build.sbt]: I probably missed it. Will take another look.

**Comment by michaelpigg** [comment]
@gkossakowski It works for me. Here is what I did:
1. Clone repo into new directory
2. checkout partest-sbt-build
3. run `sbt test/test`

I let it run through scalacheck, instrumented, and part of pos. If you're still seeing the issue, can you paste the .log files in files/test/scalacheck?

**Comment by michaelpigg** [comment]
I did notice that I haven't addressed the issue of getting the jar files populated into test/files/lib automatically. That's not the cause of the failure in scalacheck, but it will prevent other tests from being successful.

**Comment by gkossakowski** [comment]
Here's are the log files:

```
cat t4147-scalacheck.log
t4147.scala:67: error: not enough arguments for method throws: (x: => Any, c: Class[T])org.scalacheck.Prop.
Unspecified value parameter c.
    throws(classOf[NullPointerException])(mutable.TreeSet.empty[Int](null))
          ^
one error found
Greks-MBP:scalacheck grek (partest-sbt-build)$ cat treeset-scalacheck.log
treeset.scala:37: warning: This catches all Throwables. If this is really intended, use `case _ : Throwable` to clear this warning.
    try { while (it.hasNext) it.next; true } catch { case _ => false }
                                                          ^
treeset.scala:154: error: not enough arguments for method throws: (x: => Any, c: Class[T])org.scalacheck.Prop.
Unspecified value parameter c.
    throws(classOf[NullPointerException])(TreeSet.empty[Int](null))
          ^
one warning found
one error found
```

Could it be I somehow get the wrong version of scalacheck used?

**Comment by gkossakowski** [comment]
Hey! I found the problem. It turned out that I had some stale version of scalacheck.jar in `test/files/lib` directory. Once I removed that jar, partest stopped picking it up and tests started to pass. Sorry for the noise.

**Comment by michaelpigg** [comment]
Phew, glad that's solved. I have a change to get /test/files/lib populated ready to push. I'll rebase to pick up the changes that have been merged as well.

**Comment by michaelpigg** [comment]
I pushed a change that combines all previous commits and is rebased to current `sbt-build`. It runs `pull-binary-libs.sh` to ensure /test/files/lib is populated.

I bound partest to IntegrationTest now, so it is run with `sbt test/it:test`.

**Comment by gkossakowski** [comment]
LGTM!

Excellent work!

**Comment by gkossakowski** [comment]
@michaelpigg: are you at Scala Days?

**Comment by michaelpigg** [comment]
Yes :)
